### PR TITLE
[Content Addressable] Part 3: Include Ruby ABI in uniqueness checks and duplicate detection

### PR DESCRIPTION
**Pull request by OughtPuts**
### Changes:
- Updates Version uniqueness checks to include ruby_abi when comparing duplicate versions.
- Updates the push-path duplicate lookup so skinny binaries with different Ruby ABIs can coexist, while same-ABI
  duplicates are still rejected.
- Replaces the existing Version unique indexes with partial unique indexes that preserve old fat/source gem uniqueness
  when ruby_abi is NULL, and include ruby_abi for skinny gems.
- Updates schema for the new indexes.

  ### Testing:
- Tests added to version_test for duplicate version and canonical version behavior with same/different Ruby ABIs.
- Tests added to pusher_test for push duplicate detection with same/different Ruby ABIs.
- Patch `models/pusher_test` to stub `required_ruby_version` so that the new validations don't cause test failures.

To tophat:

Pull branch down, ensure migrations have been run.

```
rubygem = Rubygem.find_or_create_by!(name: "skinny-tophat-different-abi")

v1 = rubygem.versions.create!(
  number: "1.0.0",
  platform: "arm64-darwin-25",
  gem_platform: "arm64-darwin-25",
  required_ruby_version: "~> 3.3.0",
  summary: "test",
  authors: ["test"],
  sha256: Digest::SHA2.base64digest("skinny-tophat-different-abi-3.3")
)

v2 = rubygem.versions.create!(
  number: "1.0.0",
  platform: "arm64-darwin-25",
  gem_platform: "arm64-darwin-25",
  required_ruby_version: "~> 3.4.0",
  summary: "test",
  authors: ["test"],
  sha256: Digest::SHA2.base64digest("skinny-tophat-different-abi-3.4")
)

[v1.ruby_abi, v2.ruby_abi]
# => ["3.3", "3.4"]

# Duplicate same-ABI skinny version should be invalid:

dup_skinny = rubygem.versions.build(
  number: "1.0.0",
  platform: "arm64-darwin-25",
  gem_platform: "arm64-darwin-25",
  required_ruby_version: "~> 3.4.0",
  summary: "test",
  authors: ["test"],
  sha256: Digest::SHA2.base64digest("skinny-tophat-different-abi-3.4-dup")
)

dup_skinny.valid?
# => false

# Fat binaries without Ruby ABIs still follow the old uniqueness behaviour.
```

### CODE REVIEW

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    add_index :versions,
+      %i[rubygem_id number platform ruby_abi],
+      unique: true,
+      where: "ruby_abi IS NOT NULL",
+      name: "index_versions_on_rubygem_id_number_platform_abi",
+      algorithm: :concurrently
+  end
+
+  def down
+    remove_index :versions,
```
[the remark this answers is not in the document]
Comment (OughtPuts) [reply db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb:37]: Not 100% sure what's preferred on this one - I normally have a bit of a sensitivity to irreversible migrations but that's just me! Guidance appreciated :)

File: app/models/version.rb [hunk 2]
```diff
@@ -492,19 +482,42 @@ def derive_ruby_abi
...
   end
 
   def gem_platform_and_number_are_unique
-    platforms = Version.where(rubygem_id: rubygem_id, number: number, gem_platform: gem_platform).pluck(:platform)
+    platforms = Version.where(rubygem_id: rubygem_id, number: number, gem_platform: gem_platform, ruby_abi: ruby_abi).pluck(:platform)
     return if platforms.empty?
-    errors.add(:base, "A version already exists with this number and resolved platform #{platforms}")
+
+    message = if ruby_abi.present?
+                "A version already exists with this number, resolved platform, and Ruby ABI #{platforms}"
```
Comment (girachawda) [inline app/models/version.rb]: [nit] I wonder if we can do something like this, so that the platform gets printed beside the resolved platform text. Instead of next to the Ruby ABI. Optionally, we can also add the ruby api in the error message, but i don't know how valuable it would be.

```suggestion
                "A version already exists with number #{number}, platform #{platform}, and Ruby ABI #{ruby_abi}."
```

**Comment by girachawda** [comment]
Tophat ✅ 
Screenshot 2026-07-02 at 2 49 42 PM

**Review by girachawda — APPROVED** [review]

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  disable_ddl_transaction!
+
+  def up
+    remove_index :versions,
+      name: "index_versions_on_canonical_number_and_rubygem_id_and_platform",
+      algorithm: :concurrently
+
+    remove_index :versions,
+      name: "index_versions_on_rubygem_id_and_number_and_platform",
+      algorithm: :concurrently
```
Comment (jenshenny) [inline db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb]: Could we remove the indexes in a follow up after everything is deployed so we can ensure the new indexes are working well?

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      where: "ruby_abi IS NULL",
+      name: "index_versions_on_canonical_number_rubygem_id_platform_no_abi",
+      algorithm: :concurrently
+
+    add_index :versions,
+      %i[canonical_number rubygem_id platform ruby_abi],
+      unique: true,
+      where: "ruby_abi IS NOT NULL",
+      name: "index_versions_on_canonical_number_rubygem_id_platform_abi",
+      algorithm: :concurrently
```
Comment (jenshenny) [inline db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb:26]: Is there a reason why we need an index for both ruby abi and null ruby abi columns?

**Review by jenshenny — COMMENTED** [review]
Just have some questions around the db indexes!

File: app/models/version.rb (hunk 2 above)
Comment (OughtPuts) [reply app/models/version.rb]: Nice, thanks for thinking about the readability. I went back and forth a few times but it's a slightly funny one because it feels like:
- maybe it's good to be cautious about changing the existing error message too much in case users have tooling etc that depends on the exact wording of that response.
- therefore we should probably make the `ruby_abi.present?` branch match as closely as poss.
- also the user has just pushed a specific version at this point just before they get the error so it's probably a bit of overkill to tell them which number and platform this is for because they know as they've just done it 😅
- as far as I understand we feed them back `platforms` because they _do_ need to know what the resolved platforms are as these are calculated server-side.
HOWEVER 
- I've definitely been inspired to try to make the line a bit more readable, so I've switched the order around so it's a bit clearer that the `platforms` variable relates to the 'resolved platform'.

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb (hunk 3 above)
Comment (OughtPuts) [reply db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb]: Have noted it on the build doc so it doesn't get missed!

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb (hunk 4 above)
Comment (OughtPuts) [reply db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb:26]: Yeah, [Postgres doesn't consider a `NULL` value as equal to another `NULL`](https://www.postgresql.org/docs/current/indexes-unique.html), meaning that we could in theory have a scenario where the canonical_number, rubygem_id and platform were the same but this gets allowed through because `ruby_abi` is `NULL` in both cases (i.e. fat binaries).

**Comment by OughtPuts** [comment]
@jenshenny Ready again if you get a gap! TIA! ⭐️

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb (hunk 4 above)
Comment (jenshenny) [reply db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb:26]: Could we use the `NULLS NOT DISTINCT` option when creating the index so nulls are treated as equal?

File: test/models/version_test.rb
Comment (jenshenny) [inline test/models/version_test.rb:1]: Should we add a test to disallow duplicate versions with null ABIs? We should also test the db constraints are also working correctly (we are only testing the Rails validations).

File: test/integration/pusher_test.rb [hunk 5]
```diff
@@ -314,6 +314,44 @@ def two_cert_chain(signing_key:, root_not_before: Time.current, cert_not_before:
     [root_cert, cert]
   end
 
+  context "finding an existing gem version" do
```
Comment (jenshenny) [inline test/integration/pusher_test.rb]: There's a context named "finding an existing gem" in `test/models/pusher_test.rb`. I think it makes sense to move the tests there as the integration tests test the whole push flow whereas the model tests test specific sections. wdyt?

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb (hunk 3 above)
Comment (OughtPuts) [reply db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb]: On giving this a bit more thought we actually can't remove the old index after deploy because the old index (`rubygem_id`, `number`, `platform`) will block the creation of skinny gems that have those values as the same, just with a different `ruby_abi` 😞 I've just confirmed this on my local as well. I could split it into a separate migration + commit if that makes it a bit safer?

**Comment by OughtPuts** [comment]
@jenshenny Thanks for the feedback - this is ready for another look! :) Let me know how you want to handle the old index removal - I don't think we can do it after deploy but we could put it in an entirely separate commit and migration right at the end if you're happy with that.

File: test/models/version_test.rb [hunk 6]
```diff
@@ -159,6 +159,68 @@ class VersionTest < ActiveSupport::TestCase
...
+        gem_platform: "arm64-darwin-25",
+        required_ruby_version: ">= 3.2"
+      )
+
+      assert_nil duplicate.ruby_abi
+
+      assert_raises ActiveRecord::RecordNotUnique do
+        duplicate.save!(validate: false)
+      end
+    end
```
Comment (jenshenny) [inline test/models/version_test.rb]: There's a typo in the test name, + I think it's better if we test another fat `required_ruby_version` value and the Rails validation.

```suggestion
    should "not allow duplicate versions with nil Ruby ABIs" do
      create(
        :version,
        rubygem: @rubygem,
        number: "1.0.0",
        platform: "arm64-darwin-25",
        gem_platform: "arm64-darwin-25",
        required_ruby_version: ">= 3.2"
      )

      duplicate = build(
        :version,
        rubygem: @rubygem,
        number: "1.0.0",
        platform: "arm64-darwin-25",
        gem_platform: "arm64-darwin-25",
        required_ruby_version: ">= 3.0"
      )

      refute_predicate version, :valid?

      assert_raises ActiveRecord::RecordNotUnique do
        duplicate.save!(validate: false)
      end
    end
```

File: test/models/version_test.rb [hunk 7]
```diff
@@ -159,6 +159,68 @@ class VersionTest < ActiveSupport::TestCase
...
+
+    should "not allow duplicate versions with the same Ruby ABI" do
+      existing_version = create(:version, rubygem: @rubygem, number: "1.0.0", platform: "arm64-darwin-25", required_ruby_version: "~> 3.3.0")
+
+      version = build(:version, rubygem: @rubygem, number: existing_version.number, platform: existing_version.platform,
+        gem_platform: existing_version.gem_platform, required_ruby_version: "~> 3.3.0")
+
+      refute_predicate version, :valid?
+    end
```
Comment (jenshenny) [inline test/models/version_test.rb:182]: nit: We should test the db validation here too

```suggestion
      refute_predicate version, :valid?
      
      assert_raises ActiveRecord::RecordNotUnique do
        duplicate.save!(validate: false)
      end
    end
```

**Review by jenshenny — APPROVED** [review]
> Let me know how you want to handle the old index removal - I don't think we can do it after deploy but we could put it in an entirely separate commit and migration right at the end if you're happy with that.

Right, because the unique validation would be too restrictive. Let's add a commit with another migration to remove the old index  👍

File: db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb (hunk 4 above)
Comment (OughtPuts) [reply db/migrate/20260701131420_add_ruby_abi_to_version_unique_indexes.rb:26]: Ah damn, it looks like the `NULLS NOT DISTINCT` is only available for [postgres 15 and later](https://www.postgresql.org/docs/release/15.0/), and we are pinning to 14.15 according to docker-compose.yml. 

I would suggest going back to the multiple index option, would you be happy with that? Happy to be steered somewhere else if there's other considerations I'm not aware of :)
